README.md
# Another Pong
A fully-commented Pong clone written in 6502 assembly for the Atari 2600.

<p align="center">
  <img alt="another-pong" src="./media/another-pong.png" width="75%" />
</p>

## Features
* Two-player support
* Sound effects

## Build and Run
**The included `Makefile` assumes that [DASM assembler](https://github.com/dasm-assembler/dasm) and the Atari 2600 emulator [Stella](https://github.com/stella-emu/stella) are installed and available from the command line.**
* Call `make` to assemble the `another-pong.bin` ROM.
* Call `make run` to run the ROM file via Stella.
* Call `make debug` to run the ROM in Stella's debug mode.


CHANGELOG.md
# 1.0.1 (May 24, 2022)
*  2573338 Update Makefile commands and README.md

# 1.0.0 (May 24, 2022)
*  9b53283 Extract game loop, subroutines and data tables into own modules
*  b5861ee Implement sounds and use higher timer interval for overscan period
*  d3d6a60 Implement scoreboard and set ball priority to highest
*  1d79b4c Implement center net graphics, handle x-velocity reversal on goal
*  9888d7e Implement player collision handling, updating x-velocity of ball
*  49a7057 Implement screen top and bottom collision handling for Ball
*  a11cf21 Add cycle counts to each instruction, and basic x-offset updates to the Ball in overscan period
*  c3ca832 Refactor player x-position handling using fine positioning
*  d12e422 Implement basic ball drawing routines


data.asm
;===========================================================================
; Another Pong - by Elliott Fiedler (@leitdeux)
; data.asm - data tables
;===========================================================================
    align $100              ; ensure data doesn't cross page boundary

Digits:
    .byte %01110111          ; ### ###
    .byte %01010101          ; # # # #
    .byte %01010101          ; # # # #
    .byte %01010101          ; # # # #
    .byte %01110111          ; ### ###

    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #

    .byte %01110111          ; ### ###
    .byte %00010001          ;   #   #
    .byte %01110111          ; ### ###
    .byte %01000100          ; #   #
    .byte %01110111          ; ### ###

    .byte %01110111          ; ### ###
    .byte %00010001          ;   #   #
    .byte %00110011          ;  ##  ##
    .byte %00010001          ;   #   #
    .byte %01110111          ; ### ###

    .byte %01010101          ; # # # #
    .byte %01010101          ; # # # #
    .byte %01110111          ; ### ###
    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #

    .byte %01110111          ; ### ###
    .byte %01000100          ; #   #
    .byte %01110111          ; ### ###
    .byte %00010001          ;   #   #
    .byte %01110111          ; ### ###

    .byte %01110111          ; ### ###
    .byte %01000100          ; #   #
    .byte %01110111          ; ### ###
    .byte %01010101          ; # # # #
    .byte %01110111          ; ### ###

    .byte %01110111          ; ### ###
    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #
    .byte %00010001          ;   #   #

    .byte %01110111          ; ### ###
    .byte %01010101          ; # # # #
    .byte %01110111          ; ### ###
    .byte %01010101          ; # # # #
    .byte %01110111          ; ### ###

    .byte %01110111          ; ### ###
    .byte %01010101          ; # # # #
    .byte %01110111          ; ### ###
    .byte %00010001          ;   #   #
    .byte %01110111          ; ### ###



game.asm
;===========================================================================
; Another Pong - by Elliott Fiedler (@leitdeux)
; game.asm - the game loop
; - by Elliott Fiedler (@leitdeux)
;===========================================================================
NextFrame:
    lsr SWCHB               ; 5?, handle game reset
    bcc Start               ; 2

    lda #SCANLINE_COUNT     ; 2, reset visible scanline count
    sta ScanlineCount       ; 3

;===========================================================================
; Output VSync, VBlank
;===========================================================================
    lda #2                  ; 2, enable VBLANK, VSYNC
    sta VBLANK              ; 3
    sta VSYNC               ; 3

    sta WSYNC               ; 3, generate 3 lines of VSYNC
    sta WSYNC               ; 3
    sta WSYNC               ; 3

    lda #0                  ; 2, disable VSYNC
    sta VSYNC               ; 3

    ldx #37                 ; 2
OutputVBlank:
    sta WSYNC               ; 3
    dex                     ; 2
    bne OutputVBlank        ; 2
    stx VBLANK              ; 3, disable VBLANK

    jsr HandleScoreDraw     ; draw scoreboard

    ; set initial player positions after scoreboard draw
    lda Player0X            ; 3, handle P0 x-position
    ldx #0                  ; 2
    jsr HandleObjXPosition  ; 39

    lda Player1X            ; 3, handle P1 x-position
    ldx #1                  ; 2
    jsr HandleObjXPosition  ; 39

    lda #0                  ; 2, set background color
    sta COLUBK              ; 3
    lda #COLOR_YELLOW       ; 2, set ball color
    sta COLUPF              ; 3
    lda #COLOR_WHITE        ; 2, set player colors
    sta COLUP0              ; 3
    sta COLUP1              ; 3

;===========================================================================
; Visible scanlines
;===========================================================================
DrawGame:
    ; FIXME: -- improve drawing routines so that setup consistently occurs during h-blank period
    lda ScanlineCount       ; 3, load visible scanline count
VisibleScanlines:           ; 78
    ldx #0                  ; 2, draw P0
    jsr HandlePlayerDraw    ; 34
    ;sta WSYNC               ; 3, scanline (1 of 2) we sync after because we want
                            ; P0 and P1 to be drawn on the same line
    lda ScanlineCount       ; 3, restore scanline value for next player draw
    ldx #1                  ; 2, draw P1
    jsr HandlePlayerDraw    ; 34
.HandleMissileDraw:         ; draws the centered "tennis net"
    ldx #0                  ; 2, value for disabling M0 graphics
    lda #4                  ; 2, the vertical width of M0 when AND-ed with ScanlineCount
    bit ScanlineCount
    beq .NoMissileDraw      ; 2, if the result of AND is not 0, do not draw M0
    ldx #2                  ; 2, enable M0 graphics
.NoMissileDraw:
    stx ENAM0               ; 3, enable or disable M0 graphics
    lda ScanlineCount       ; 3, load current scanline
.HandleBallDraw:
    ldx #BALL_SIZE          ; 2, load ball width in X
    ldy #%00000010          ; 2, load ball-graphics-enable value in Y
    sec                     ; 2, set carry flag before subtraction
    sbc BallY               ; 3, subtract ball y-position from current scanline count
    cmp #BALL_HEIGHT        ; 2, is ball found in scanline?
    bcc .DrawBall           ; 2

    ldx #0                  ; 2, draw nothing
    ldy #0                  ; 2
.DrawBall:
    sta WSYNC               ; 3, draw scanline (2 of 2)
    stx CTRLPF              ; 3, set ball draw width
    sty ENABL               ; 3, enable ball draw

    dec ScanlineCount       ; 5, decrement visible scanline count
    lda ScanlineCount       ; 3, load updated scanline count
    bne VisibleScanlines    ; 2, branch if all scanlines have been drawn

;===========================================================================
; Output overscan
; - ensure we spend only 29 scanlines worth of work in here
; - at the end of it all, we call again WSYNC and HMOVE to apply motion updates
;===========================================================================
    ;lda #2                  ; 2, enable VBLANK
    ;sta VBLANK              ; 3
    lda #COLOR_BLUE
    sta COLUBK
OutputOverscan:
    ; TIMER_SETUP 29
    TIMER_SETUP 40           ; use timer instead of iterating over each scanline

;===========================================================================
; Handle Ball collisions with screen (update ball y-velocity)
; X - ball y-position
; Y - audio control
; A - ball y-velocity (+1 or -1)
;===========================================================================
HandleScreenCollision:
    ldx BallY               ; load ball y-position in X
    ldy #%00001011
    lda #%11111000
    sta AUDF0
    lda #$ff                ; 2, load negative y-velocity (-1) in A
    cpx #SCREEN_Y_MAX       ; is ball at top of screen?
    bcs .UpdateBallYVel
    lda #1                  ; 2, load positive y-velocity (1) in A
    cpx #SCREEN_Y_MIN       ; is ball at bottom of screen?
    bcc .UpdateBallYVel
    ldy #0
    lda BallYVel            ; load current y-velocity value
.UpdateBallYVel:
    sta BallYVel            ; set ball y-velocity to either -1, 1 or its current value
    cpy #0
    bne .PlayAudio
.PlayAudio:
    sty AUDC0
    sty AUDV0

;===========================================================================
; Handle Ball collision with "goal" (screen left, right)
; - compare the current frame count with the target frame count
; - if GoalFrameCount >= GoalFrameTarget, then consider the ball has reached
; the "goal" point; i.e. a player has scored a point
; - reset the fine x-position of the ball
; - reset the frame count and frame target
; - reverse the direction of the ball (x-velocity)
; - update the score of the player that earned a point
;===========================================================================
HandleGoalCollision:
    lda GoalFrameCount      ; 3, load current frame count in A
    cmp GoalFrameTarget     ; 3
    bcc .NoGoal             ; 2, is frame count < max frames?
    lda #BALL_Y_INIT        ; 2, reset ball y-position
    sta BallY
    lda BallX               ; 3, reset ball x-position
    ldx #4                  ; 2
    jsr HandleObjXPosition  ; 39
    lda #0                  ; reset frame count
    sta GoalFrameCount
    lda #CENTER_TO_GOAL_FRAMES ; reset target # of frames until goal should be checked
    sta GoalFrameTarget
    ldy #$30                ; if x-velocity is negative, set positive and update P1 score
    ldx #1
    lda BallXVel
    bmi .ReflectXVel
    ldy #$d0                ; if x-velocity is positive, set negative and update P0 score
    ldx #0
.ReflectXVel:
    sty BallXVel
    sed                     ; 2, enable BCD
    lda Player0Score,x      ; 4
    clc
    adc #1                  ; 2, increment score
    sta Player0Score,x      ; 4
    cld                     ; 2, disable BCD
    jmp .CommitBallUpdate
.NoGoal:
    inc GoalFrameCount      ; 5, increment frame count

;===========================================================================
; Handle Ball collisions with players (update ball x-velocity)
; - if collides with P1, set x-velocity to negative
; - if collides with P0, set x-velocity to positive
;===========================================================================
HandlePlayerCollision:
    ldy #0
    lda BallXVel            ; 3, load current x-velocity in A
    bmi .HandleBallMoveLeft ; 2, is ball moving left?
    lda #%01000000          ; 2, has ball collided with P1?
    bit CXP1FB              ; 3
    bvc .NoPlayerCollision  ; 2
    ldy #%00011101
    lda #$d0                ; 2, update x-velocity to negative value
    sta BallXVel            ; 3
    lda #0                  ; 2, reset goal frame counter and update target frames
    sta GoalFrameCount
    lda #PLAYER_TO_GOAL_FRAMES
    sta GoalFrameTarget
.HandleBallMoveLeft:
    lda #%01000000          ; 2, has ball collided with P0?
    bit CXP0FB              ; 3
    bvc .NoPlayerCollision  ; 2
    ldy #%00011101
    lda #$30                ; 2, update x-velocity to positive value
    sta BallXVel            ; 3
    lda #0                  ; 2, reset goal frame counter
    sta GoalFrameCount
    lda #PLAYER_TO_GOAL_FRAMES
    sta GoalFrameTarget
.NoPlayerCollision:
    sty AUDC0
    sty AUDV0
    sta CXCLR               ; 3, clear collision registers

;===========================================================================
; Update Ball x,y position via velocity
;===========================================================================
UpdateBallPosition:
    sta WSYNC               ; 3, necessary?
    sta HMCLR               ; 3

    lda BallY               ; 3, update y-position
    clc                     ; 2, clear carry
    adc BallYVel            ; 3, increment y-position by y-velocity
    sta BallY               ; 3
    lda BallXVel            ; 3, update x-position
    bmi .MoveBallLeft       ; 2, if x-velocity < 0, move Ball left

    clc                     ; 2, clear carry
    adc BallXFrac           ; 3, increment x-position by x-velocity
    sta BallXFrac           ; 3, BallXFrac += BallXVel
    lda #$f0                ; 2, set ball motion register to move right 1 clock
    sta HMBL                ; 3
    bne .CommitBallUpdate   ; 2

.MoveBallLeft:
    clc                     ; 2, set carry flag
    adc BallXFrac           ; seems to work...? looks strange, however
    sta BallXFrac           ; 3, update x-velocity value
    lda #$10                ; 2, set ball motion register to move left by 1 clock
    sta HMBL                ; 3
.CommitBallUpdate:
    sta WSYNC               ; 3
    sta HMOVE               ; 3, commit all horizontal motion updates

;===========================================================================
; Handle player input for next frame (88)
;===========================================================================
HandlePlayerInputs:
    lda #%00010000          ; 2, handle P0 input
    ldx Player0Y            ; 3, load P0 y-position
    jsr HandlePlayerInput   ; 36
    stx Player0Y            ; 3, update P0 y-position

    lda #%00000001          ; 2, handle P1 input
    ldx Player1Y            ; 3, load P1 y-position
    jsr HandlePlayerInput   ; 36
    stx Player1Y            ; 3, update P1 y-position

    jsr ParseScoreDigits    ; update score digits

;-----------------------------------------------------------------------
; Overscan timer expires
;-----------------------------------------------------------------------
    TIMER_WAIT              ; wait until 29 scanlines, and then strobe
                            ; one more scanline for 30 total in overscan
    jmp NextFrame



main.asm
;===========================================================================
; Another Pong - by Elliott Fiedler (@leitdeux)
; main.asm
;===========================================================================
	processor 6502

    .include "includes/vcs.h"
    .include "includes/macro.h"
    .include "includes/timer.h"

;===========================================================================
; Declare variables starting from memory address $80
;===========================================================================
    seg.u Variables
    org $80

;-----------------------------------------------------------------------
; Variables
;-----------------------------------------------------------------------
ScanlineCount .byte         ; number of visible scanlines

Player0X .byte              ; P0 x-position
Player1X .byte              ; P1 x-position
Player0Y .byte              ; P0 y-position
Player1Y .byte              ; P1 y-position

BallX .byte                 ; ball x-position
BallY .byte                 ; ball y-position
BallXVel .byte              ; ball x velocity (signed)
BallYVel .byte              ; ball y velocity (signed)
BallXFrac .byte             ; ball x fractional value
                            ; BallXVel is added to BallXFrac so that the Ball
                            ; can move in incremenets less than 1px per frame

GoalFrameCount .byte        ; the current number of frames elapsed since a goal was made
GoalFrameTarget .byte       ; max # of frames until reaching a goal

Player0Score .byte          ; score of P0, 2 digits stored as binary-coded decimal (BCD)
Player1Score .byte          ; score of P1, 2 digits (BCD)
ScoreTemp .byte             ; temporary value used in score parsing

Player0ScoreSprite .byte    ; score sprite of P0
Player1ScoreSprite .byte    ; score sprite of P1

OnesDigitOffset .word       ; score digit pointers
TensDigitOffset .word

;-----------------------------------------------------------------------
; Constants
;-----------------------------------------------------------------------
SCOREBOARD_HEIGHT = 14
SCANLINE_COUNT = 96 - SCOREBOARD_HEIGHT         ; 2-line kernel; 192 / 2 = 96
SCREEN_X_MIN = 0            ; optical left edge of screen
SCREEN_X_MAX = 160          ; optical right edge of screen
SCREEN_Y_MAX = 80           ; optical top of the screen
SCREEN_Y_MIN = 4            ; optical bottom of the screen

PLAYER_HEIGHT = 16
PLAYER_0_X_INIT = 8         ; initial P0 x-position (in pixels)
PLAYER_1_X_INIT = 152       ; initial P1 x-position
PLAYER_Y_INIT = 35          ; initial Player y-position
PLAYER_Y_MAX = 65           ; max player y-position
PLAYER_SPRITE = %00000111

BALL_HEIGHT = 4
BALL_X_INIT = 88            ; initial Ball x-position (offset)
BALL_Y_INIT = 110           ; initial Ball y-position
BALL_SIZE = %00101100       ; set 3-pixel width and highest draw priority

CENTER_TO_GOAL_FRAMES = 77  ; # of frames from screen center to a goal
PLAYER_TO_GOAL_FRAMES = 143 ; # of frames from one player to a goal

SCORE_DIGITS_HEIGHT = 5     ; the height of player score digits

COLOR_WHITE = $0e
COLOR_BLUE = $a0
COLOR_PURPLE = $8e
COLOR_YELLOW = $1a
COLOR_GREEN = $c0

;===========================================================================
; Init code segment
;===========================================================================
    seg Code
    org $f000

Start:
    CLEAN_START

;-----------------------------------------------------------------------
; Initialize variables
;-----------------------------------------------------------------------
    lda #PLAYER_Y_INIT      ; 2, set default y-position
    sta Player0Y            ; 3
    sta Player1Y            ; 3

    lda #PLAYER_0_X_INIT    ; 2, set player default x-position
    sta Player0X            ; 3
    lda #PLAYER_1_X_INIT    ; 2
    sta Player1X            ; 3

    lda #BALL_Y_INIT        ; 2, set default ball y-position
    sta BallY               ; 3

    lda #BALL_X_INIT        ; 2, set ball default x-position
    sta BallX               ; 3

    lda #1                  ; 2, set default ball velocity
    sta BallYVel            ; 3
    lda #$30                ; 2, set ball x-velocity to 48
    sta BallXVel

    lda #0                  ; 2, init goal frame count
    sta GoalFrameCount
    lda #CENTER_TO_GOAL_FRAMES ; 2, set current target frame count until goal
    sta GoalFrameTarget

;-----------------------------------------------------------------------
; Initialize fine x-position offsets for all motion objects (144)
;-----------------------------------------------------------------------
    sta WSYNC               ; 3, clear horiz. motion registers
    sta HMCLR               ; 3

    lda BallX               ; 3, handle Ball x-position
    ldx #4                  ; 2
    jsr HandleObjXPosition  ; 39

    lda #BALL_X_INIT        ; 2, handle M0 x-position
    ldx #2                  ; 2
    jsr HandleObjXPosition  ; 39

    jsr ParseScoreDigits    ; setup score digits

    sta WSYNC               ; 3, sync scanline before setting horizontal registers
    sta HMOVE               ; 3, apply *all* horizontal motion registers (HM__)

;===========================================================================
; Game loop, subroutines and data tables 
;===========================================================================
    .include "game.asm"
    .include "helpers.asm"
    .include "data.asm"

;===========================================================================
; Complete ROM
;===========================================================================
    org $fffc       ; move to memory address $fffc
    .word Start      ; write 2 bytes with program reset address
    .word Start      ; write 2 butes with interruption vector


include/timer.h
; https://8bitworkshop.com/v3.6.0/?file=examples%2Fcollisions.a&platform=vcs#
;-------------------------------------------------------
; Usage: TIMER_SETUP lines
; where lines is the number of scanlines to skip (> 2).
; The timer will be set so that it expires before this number
; of scanlines. A WSYNC will be done first.

    MAC TIMER_SETUP
.lines  SET {1}
.cycles SET ((.lines * 76) - 13)
; special case for when we have two timer events in a line
; and our 2nd event straddles the WSYNC boundary
	if (.cycles % 64) < 12
		lda #(.cycles / 64) - 1
		sta WSYNC
        else
		lda #(.cycles / 64)
		sta WSYNC
        endif
        sta TIM64T
    ENDM

;-------------------------------------------------------
; Use with TIMER_SETUP to wait for timer to complete.
; Performs a WSYNC afterwards.

    MAC TIMER_WAIT
.waittimer
        lda INTIM
        bne .waittimer
        sta WSYNC
    ENDM


includes/macro.h
; MACRO.H
; Version 1.06, 3/SEPTEMBER/2004

VERSION_MACRO         = 106

;
; THIS FILE IS EXPLICITLY SUPPORTED AS A DASM-PREFERRED COMPANION FILE
; PLEASE DO *NOT* REDISTRIBUTE MODIFIED VERSIONS OF THIS FILE!
;
; This file defines DASM macros useful for development for the Atari 2600.
; It is distributed as a companion machine-specific support package
; for the DASM compiler. Updates to this file, DASM, and associated tools are
; available at at http://www.atari2600.org/dasm
;
; Many thanks to the people who have contributed.  If you take issue with the
; contents, or would like to add something, please write to me
; (atari2600@taswegian.com) with your contribution.
;
; Latest Revisions...
;
; 1.06  03/SEP/2004     - nice revision of VERTICAL_BLANK (Edwin Blink)
; 1.05  14/NOV/2003     - Added VERSION_MACRO equate (which will reflect 100x version #)
;                         This will allow conditional code to verify MACRO.H being
;                         used for code assembly.
; 1.04  13/NOV/2003     - SET_POINTER macro added (16-bit address load)
;
; 1.03  23/JUN/2003     - CLEAN_START macro added - clears TIA, RAM, registers
;
; 1.02  14/JUN/2003     - VERTICAL_SYNC macro added
;                         (standardised macro for vertical synch code)
; 1.01  22/MAR/2003     - SLEEP macro added.
;                       - NO_ILLEGAL_OPCODES switch implemented
; 1.0	22/MAR/2003		Initial release

; Note: These macros use illegal opcodes.  To disable illegal opcode usage,
;   define the symbol NO_ILLEGAL_OPCODES (-DNO_ILLEGAL_OPCODES=1 on command-line).
;   If you do not allow illegal opcode usage, you must include this file
;   *after* including VCS.H (as the non-illegal opcodes access hardware
;   registers and require them to be defined first).

; Available macros...
;   SLEEP n             - sleep for n cycles
;   VERTICAL_SYNC       - correct 3 scanline vertical synch code
;   CLEAN_START         - set machine to known state on startup
;   SET_POINTER         - load a 16-bit absolute to a 16-bit variable

;-------------------------------------------------------------------------------
; SLEEP duration
; Original author: Thomas Jentzsch
; Inserts code which takes the specified number of cycles to execute.  This is
; useful for code where precise timing is required.
; ILLEGAL-OPCODE VERSION DOES NOT AFFECT FLAGS OR REGISTERS.
; LEGAL OPCODE VERSION MAY AFFECT FLAGS
; Uses illegal opcode (DASM 2.20.01 onwards).

            MAC SLEEP            ;usage: SLEEP n (n>1)
.CYCLES     SET {1}

                IF .CYCLES < 2
                    ECHO "MACRO ERROR: 'SLEEP': Duration must be > 1"
                    ERR
                ENDIF

                IF .CYCLES & 1
                    IFNCONST NO_ILLEGAL_OPCODES
                        nop 0
                    ELSE
                        bit VSYNC
                    ENDIF
.CYCLES             SET .CYCLES - 3
                ENDIF

                REPEAT .CYCLES / 2
                    nop
                REPEND
            ENDM

;-------------------------------------------------------------------------------
; VERTICAL_SYNC
; revised version by Edwin Blink -- saves bytes!
; Inserts the code required for a proper 3 scanline vertical sync sequence
; Note: Alters the accumulator

; OUT: A = 0

             MAC VERTICAL_SYNC
                lda #%1110          ; each '1' bits generate a VSYNC ON line (bits 1..3)
.VSLP1          sta WSYNC           ; 1st '0' bit resets Vsync, 2nd '0' bit exit loop
                sta VSYNC
                lsr
                bne .VSLP1          ; branch until VYSNC has been reset
             ENDM

;-------------------------------------------------------------------------------
; CLEAN_START
; Original author: Andrew Davie
; Standardised start-up code, clears stack, all TIA registers and RAM to 0
; Sets stack pointer to $FF, and all registers to 0
; Sets decimal mode off, sets interrupt flag (kind of un-necessary)
; Use as very first section of code on boot (ie: at reset)
; Code written to minimise total ROM usage - uses weird 6502 knowledge :)

            MAC CLEAN_START
                sei
                cld

                ldx #0
                txa
                tay
.CLEAR_STACK    dex
                txs
                pha
                bne .CLEAR_STACK     ; SP=$FF, X = A = Y = 0

            ENDM

;-------------------------------------------------------
; SET_POINTER
; Original author: Manuel Rotschkar
;
; Sets a 2 byte RAM pointer to an absolute address.
;
; Usage: SET_POINTER pointer, address
; Example: SET_POINTER SpritePTR, SpriteData
;
; Note: Alters the accumulator, NZ flags
; IN 1: 2 byte RAM location reserved for pointer
; IN 2: absolute address

            MAC SET_POINTER
.POINTER    SET {1}
.ADDRESS    SET {2}

                LDA #<.ADDRESS  ; Get Lowbyte of Address
                STA .POINTER    ; Store in pointer
                LDA #>.ADDRESS  ; Get Hibyte of Address
                STA .POINTER+1  ; Store in pointer+1

            ENDM

;-------------------------------------------------------
; BOUNDARY byte#
; Original author: Denis Debro (borrowed from Bob Smith / Thomas)
;
; Push data to a certain position inside a page and keep count of how
; many free bytes the programmer will have.
;
; eg: BOUNDARY 5    ; position at byte #5 in page

.FREE_BYTES SET 0
   MAC BOUNDARY
      REPEAT 256
         IF <. % {1} = 0
            MEXIT
         ELSE
.FREE_BYTES SET .FREE_BYTES + 1
            .byte $00
         ENDIF
      REPEND
   ENDM


; EOF

includes/vcs.h
; VCS.H
; Version 1.05, 13/November/2003

VERSION_VCS         = 105

; THIS IS A PRELIMINARY RELEASE OF *THE* "STANDARD" VCS.H
; THIS FILE IS EXPLICITLY SUPPORTED AS A DASM-PREFERRED COMPANION FILE
; PLEASE DO *NOT* REDISTRIBUTE THIS FILE!
;
; This file defines hardware registers and memory mapping for the
; Atari 2600. It is distributed as a companion machine-specific support package
; for the DASM compiler. Updates to this file, DASM, and associated tools are
; available at at http://www.atari2600.org/dasm
;
; Many thanks to the original author(s) of this file, and to everyone who has
; contributed to understanding the Atari 2600.  If you take issue with the
; contents, or naming of registers, please write to me (atari2600@taswegian.com)
; with your views.  Please contribute, if you think you can improve this
; file!
;
; Latest Revisions...
; 1.05  13/NOV/2003      - Correction to 1.04 - now functions as requested by MR.
;                        - Added VERSION_VCS equate (which will reflect 100x version #)
;                          This will allow conditional code to verify VCS.H being
;                          used for code assembly.
; 1.04  12/NOV/2003     Added TIA_BASE_WRITE_ADDRESS and TIA_BASE_READ_ADDRESS for
;                       convenient disassembly/reassembly compatibility for hardware
;                       mirrored reading/writing differences.  This is more a
;                       readability issue, and binary compatibility with disassembled
;                       and reassembled sources.  Per Manuel Rotschkar's suggestion.
; 1.03  12/MAY/2003     Added SEG segment at end of file to fix old-code compatibility
;                       which was broken by the use of segments in this file, as
;                       reported by Manuel Polik on [stella] 11/MAY/2003
; 1.02  22/MAR/2003     Added TIMINT($285)
; 1.01	        		Constant offset added to allow use for 3F-style bankswitching
;						 - define TIA_BASE_ADDRESS as $40 for Tigervision carts, otherwise
;						   it is safe to leave it undefined, and the base address will
;						   be set to 0.  Thanks to Eckhard Stolberg for the suggestion.
;                          Note, may use -DLABEL=EXPRESSION to define TIA_BASE_ADDRESS
;                        - register definitions are now generated through assignment
;                          in uninitialised segments.  This allows a changeable base
;                          address architecture.
; 1.0	22/MAR/2003		Initial release


;-------------------------------------------------------------------------------

; TIA_BASE_ADDRESS
; The TIA_BASE_ADDRESS defines the base address of access to TIA registers.
; Normally 0, the base address should (externally, before including this file)
; be set to $40 when creating 3F-bankswitched (and other?) cartridges.
; The reason is that this bankswitching scheme treats any access to locations
; < $40 as a bankswitch.

			IFNCONST TIA_BASE_ADDRESS
TIA_BASE_ADDRESS	= 0
			ENDIF

; Note: The address may be defined on the command-line using the -D switch, eg:
; dasm.exe code.asm -DTIA_BASE_ADDRESS=$40 -f3 -v5 -ocode.bin
; *OR* by declaring the label before including this file, eg:
; TIA_BASE_ADDRESS = $40
;   include "vcs.h"

; Alternate read/write address capability - allows for some disassembly compatibility
; usage ; to allow reassembly to binary perfect copies).  This is essentially catering
; for the mirrored ROM hardware registers.

; Usage: As per above, define the TIA_BASE_READ_ADDRESS and/or TIA_BASE_WRITE_ADDRESS
; using the -D command-line switch, as required.  If the addresses are not defined,
; they default to the TIA_BASE_ADDRESS.

     IFNCONST TIA_BASE_READ_ADDRESS
TIA_BASE_READ_ADDRESS = TIA_BASE_ADDRESS
     ENDIF

     IFNCONST TIA_BASE_WRITE_ADDRESS
TIA_BASE_WRITE_ADDRESS = TIA_BASE_ADDRESS
     ENDIF

;-------------------------------------------------------------------------------

			SEG.U TIA_REGISTERS_WRITE
			ORG TIA_BASE_WRITE_ADDRESS

	; DO NOT CHANGE THE RELATIVE ORDERING OF REGISTERS!

VSYNC       ds 1    ; $00   0000 00x0   Vertical Sync Set-Clear
VBLANK		  ds 1	  ; $01   xx00 00x0   Vertical Blank Set-Clear
WSYNC		    ds 1	  ; $02   ---- ----   Wait for Horizontal Blank
RSYNC			  ds 1	  ; $03   ---- ----   Reset Horizontal Sync Counter
NUSIZ0		  ds 1	  ; $04   00xx 0xxx   Number-Size player/missle 0
NUSIZ1		  ds 1	  ; $05   00xx 0xxx   Number-Size player/missle 1
COLUP0		  ds 1	  ; $06   xxxx xxx0   Color-Luminance Player 0
COLUP1      ds 1    ; $07   xxxx xxx0   Color-Luminance Player 1
COLUPF      ds 1    ; $08   xxxx xxx0   Color-Luminance Playfield
COLUBK      ds 1    ; $09   xxxx xxx0   Color-Luminance Background
CTRLPF      ds 1    ; $0A   00xx 0xxx   Control Playfield, Ball, Collisions
REFP0       ds 1    ; $0B   0000 x000   Reflection Player 0
REFP1       ds 1    ; $0C   0000 x000   Reflection Player 1
PF0         ds 1    ; $0D   xxxx 0000   Playfield Register Byte 0
PF1         ds 1    ; $0E   xxxx xxxx   Playfield Register Byte 1
PF2         ds 1    ; $0F   xxxx xxxx   Playfield Register Byte 2
RESP0       ds 1    ; $10   ---- ----   Reset Player 0
RESP1       ds 1    ; $11   ---- ----   Reset Player 1
RESM0       ds 1    ; $12   ---- ----   Reset Missle 0
RESM1       ds 1    ; $13   ---- ----   Reset Missle 1
RESBL       ds 1    ; $14   ---- ----   Reset Ball
AUDC0       ds 1    ; $15   0000 xxxx   Audio Control 0
AUDC1       ds 1    ; $16   0000 xxxx   Audio Control 1
AUDF0       ds 1    ; $17   000x xxxx   Audio Frequency 0
AUDF1       ds 1    ; $18   000x xxxx   Audio Frequency 1
AUDV0       ds 1    ; $19   0000 xxxx   Audio Volume 0
AUDV1       ds 1    ; $1A   0000 xxxx   Audio Volume 1
GRP0        ds 1    ; $1B   xxxx xxxx   Graphics Register Player 0
GRP1        ds 1    ; $1C   xxxx xxxx   Graphics Register Player 1
ENAM0       ds 1    ; $1D   0000 00x0   Graphics Enable Missle 0
ENAM1       ds 1    ; $1E   0000 00x0   Graphics Enable Missle 1
ENABL       ds 1    ; $1F   0000 00x0   Graphics Enable Ball
HMP0        ds 1    ; $20   xxxx 0000   Horizontal Motion Player 0
HMP1        ds 1    ; $21   xxxx 0000   Horizontal Motion Player 1
HMM0        ds 1    ; $22   xxxx 0000   Horizontal Motion Missle 0
HMM1        ds 1    ; $23   xxxx 0000   Horizontal Motion Missle 1
HMBL        ds 1    ; $24   xxxx 0000   Horizontal Motion Ball
VDELP0      ds 1    ; $25   0000 000x   Vertical Delay Player 0
VDELP1      ds 1    ; $26   0000 000x   Vertical Delay Player 1
VDELBL      ds 1    ; $27   0000 000x   Vertical Delay Ball
RESMP0      ds 1    ; $28   0000 00x0   Reset Missle 0 to Player 0
RESMP1      ds 1    ; $29   0000 00x0   Reset Missle 1 to Player 1
HMOVE       ds 1    ; $2A   ---- ----   Apply Horizontal Motion
HMCLR       ds 1    ; $2B   ---- ----   Clear Horizontal Move Registers
CXCLR       ds 1    ; $2C   ---- ----   Clear Collision Latches

;-------------------------------------------------------------------------------

			SEG.U TIA_REGISTERS_READ
			ORG TIA_BASE_READ_ADDRESS

                    ;											bit 7   bit 6
CXM0P       ds 1    ; $00       xx00 0000       Read Collision  M0-P1   M0-P0
CXM1P       ds 1    ; $01       xx00 0000                       M1-P0   M1-P1
CXP0FB      ds 1    ; $02       xx00 0000                       P0-PF   P0-BL
CXP1FB      ds 1    ; $03       xx00 0000                       P1-PF   P1-BL
CXM0FB      ds 1    ; $04       xx00 0000                       M0-PF   M0-BL
CXM1FB      ds 1    ; $05       xx00 0000                       M1-PF   M1-BL
CXBLPF      ds 1    ; $06       x000 0000                       BL-PF   -----
CXPPMM      ds 1    ; $07       xx00 0000                       P0-P1   M0-M1
INPT0       ds 1    ; $08       x000 0000       Read Pot Port 0
INPT1       ds 1    ; $09       x000 0000       Read Pot Port 1
INPT2       ds 1    ; $0A       x000 0000       Read Pot Port 2
INPT3       ds 1    ; $0B       x000 0000       Read Pot Port 3
INPT4       ds 1    ; $0C		x000 0000       Read Input (Trigger) 0
INPT5       ds 1	; $0D		x000 0000       Read Input (Trigger) 1

;-------------------------------------------------------------------------------

			SEG.U RIOT
			ORG $280

	; RIOT MEMORY MAP

SWCHA       ds 1    ; $280      Port A data register for joysticks:
					;			Bits 4-7 for player 1.  Bits 0-3 for player 2.

SWACNT      ds 1    ; $281      Port A data direction register (DDR)
SWCHB       ds 1    ; $282		Port B data (console switches)
SWBCNT      ds 1    ; $283      Port B DDR
INTIM       ds 1    ; $284		Timer output

TIMINT  	ds 1	; $285

		; Unused/undefined registers ($285-$294)

			ds 1	; $286
			ds 1	; $287
			ds 1	; $288
			ds 1	; $289
			ds 1	; $28A
			ds 1	; $28B
			ds 1	; $28C
			ds 1	; $28D
			ds 1	; $28E
			ds 1	; $28F
			ds 1	; $290
			ds 1	; $291
			ds 1	; $292
			ds 1	; $293

TIM1T       ds 1    ; $294		  set 1 clock interval
TIM8T       ds 1    ; $295      set 8 clock interval
TIM64T      ds 1    ; $296      set 64 clock interval
T1024T      ds 1    ; $297      set 1024 clock interval

;-------------------------------------------------------------------------------
; The following required for back-compatibility with code which does not use
; segments.

            SEG

; EOF